### PR TITLE
version 1.0

### PR DESCRIPTION
**Pull request by alexyarats**
RaccooChat Version 1.0.
Was realized multithread server. Was implemented function that makes a query every second on the server and gets all the new messages. All messages are stored in an array. And all users and positions of the last message that this user is got, also are stored in an array.

And also was implemented these functions:
- get all command that we can use in the chat (/command);
- connect and disconnect users;
- it is possible to look who is online now (/users);
- get last five message from the chat (/last).

### CODE REVIEW

File: PythonClient.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+from thrift import Thrift
+from thrift.transport import TSocket
+from thrift.transport import TTransport
+from thrift.protocol import TBinaryProtocol
+
+def printMessages(messageArray):
+    for msg in messageArray:
+        shift = 10 - len(msg.user_name)
+        print '[' + msg.time + ']' + (' ' * shift) + msg.user_name + ' : ' + msg.text
```
Comment (denisyarats) [inline PythonClient.py]: use formatting here, like print '%s ... ' % (...)

File: PythonClient.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+def printUsers(usersArray):
+    print 'Users online: ' + ', '.join(map(str, usersArray))
```
Comment (denisyarats) [inline PythonClient.py]: same here

File: PythonClient.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+def getMessages(client, user):
+    while True:
+        time.sleep(1)
+        sys.stdout.write('\r' + ' ' * (len(readline.get_line_buffer()) + 2) + '\r')
+        printMessages(client.getNewMessages(user))
+        sys.stdout.write(readline.get_line_buffer())
+        sys.stdout.flush()
+
+def main():
+    transport = TSocket.TSocket('localhost', 9090)
```
Comment (denisyarats) [inline PythonClient.py]: create a separate class for the client and put it in a separate file. develop it's interface.

Comment (denisyarats) [inline PythonClient.py]: pass host and port as binary arguments, you can use argparse for this https://docs.python.org/3/library/argparse.html

File: PythonClient.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    transport = TTransport.TBufferedTransport(transport)
+    protocol = TBinaryProtocol.TBinaryProtocol(transport)
+    client = RacconChat.Client(protocol)
+
+    transport.open()
+
+    while (True):
+        user_nickname = raw_input("Input your nickname: ")
```
Comment (denisyarats) [inline PythonClient.py]: please follow google python style guide: https://google.github.io/styleguide/pyguide.html

File: gen-cpp/CppRacconChat.h [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#include "RacconChat.h"
```
Comment (denisyarats) [inline gen-cpp/CppRacconChat.h]: move these files out of gen-cpp. Give it a better name, it shouldn't have Cpp in it. maybe RacoonChatServer or something

Comment (denisyarats) [inline gen-cpp/CppRacconChat.h]: use #pragma once: read more about it here https://en.wikipedia.org/wiki/Pragma_once

File: gen-cpp/CppRacconChat.h [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+using namespace ::tutorial;
+
+class RacconChatHandler : virtual public RacconChatIf {
+ private:
```
Comment (denisyarats) [inline gen-cpp/CppRacconChat.h]: usually you define things in the following order:
public:
proteched:
private:

File: gen-cpp/CppRacconChat.h [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  std::map<std::string, int32_t> users_online_;
+  std::vector<tutorial::Message> messages_;
+
+ public:
+  RacconChatHandler();
+  bool connectUser(const std::string&);
+  void disconnectUser(const std::string&);
+  void getAllOnlineUsers(std::map<std::string, int32_t>&);
+  void getNewMessages(std::vector<tutorial::Message>&, const std::string&);
+  void getLastFiveMessages(std::vector<tutorial::Message>&);
```
Comment (denisyarats) [inline gen-cpp/CppRacconChat.h]: use const for methods if possible, i.e. int f() const;

File: gen-cpp/CppServer.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
+#include <thrift/protocol/TBinaryProtocol.h>
```
Comment (denisyarats) [inline gen-cpp/CppServer.cpp]: the file with the main function, you can call it just server.cpp

File: gen-cpp/CppServer.cpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+#include <thrift/server/TServer.h>
```
Comment (denisyarats) [inline gen-cpp/CppServer.cpp]: use include order like this:

`#include <files from your library>`
...
`<empty line>`
`#include <files from the standard library>`
...
`<empty line>`
`#inlcude <files from the other libraries>`
...

File: gen-cpp/CppServer.cpp [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+using namespace ::apache::thrift::transport;
+using namespace ::apache::thrift::server;
+using namespace ::apache::thrift::concurrency;
+
+using boost::shared_ptr;
+
+using namespace ::tutorial;
+
+int32_t main(int32_t argc, char **argv) {
+  int32_t port = 9090;
```
Comment (denisyarats) [inline gen-cpp/CppServer.cpp]: use google gflags and glog for printing and passing parameters https://github.com/gflags/gflags

File: gen-cpp/MakeFile.md [hunk 11]
```diff
@@ -0,0 +1,2 @@
+all:
```
Comment (denisyarats) [inline gen-cpp/MakeFile.md:1]: make file usually have name Makefile.

Also we need to start using cmake: https://cmake.org/

File: gen-py/tutorial/RacconChat-remote [hunk 12]
```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python
```
Comment (denisyarats) [inline gen-py/tutorial/RacconChat-remote]: if you don't use this file, feel free to delete it

File: tutorial.thrift [hunk 13]
```diff
@@ -0,0 +1,30 @@
+#!/usr/local/bin/thrift --gen cpp
+
+namespace cpp tutorial
```
Comment (denisyarats) [inline tutorial.thrift]: get a better name for the namespace, maybe raccoochat

File: tutorial.thrift [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+namespace py tutorial
+
+typedef i32 MyInteger
```
Comment (denisyarats) [inline tutorial.thrift]: don't need this typedef, remove

File: tutorial.thrift [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+const map<string, string> MAP_COMMAND = {'users': '/users - returns all users that are online.',
```
Comment (denisyarats) [inline tutorial.thrift]: format this map:
map = {
  k: v,
  k: v,
}

File: tutorial.thrift [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+                                         'last': '/last - returns last five messages',
+                                         'exit': '/exit - sing out from the chat'}
+
+struct Message {
+  1: string time,
```
Comment (denisyarats) [inline tutorial.thrift]: use c++ style guide here

**Review by denisyarats — CHANGES_REQUESTED** [review]

File: PythonClient.py (hunk 1 above)
Comment (alexyarats) [reply PythonClient.py]: done

File: PythonClient.py (hunk 2 above)
Comment (alexyarats) [reply PythonClient.py]: done

File: PythonClient.py (hunk 4 above)
Comment (alexyarats) [reply PythonClient.py]: done

File: gen-cpp/CppRacconChat.h (hunk 5 above)
Comment (alexyarats) [reply gen-cpp/CppRacconChat.h]: the same error
image

Comment (alexyarats) [reply gen-cpp/CppRacconChat.h]: done

File: gen-cpp/CppRacconChat.h (hunk 6 above)
Comment (alexyarats) [reply gen-cpp/CppRacconChat.h]: done

File: gen-cpp/CppServer.cpp (hunk 8 above)
Comment (alexyarats) [reply gen-cpp/CppServer.cpp]: done

File: gen-cpp/CppServer.cpp (hunk 9 above)
Comment (alexyarats) [reply gen-cpp/CppServer.cpp]: what do you mean?

File: gen-py/tutorial/RacconChat-remote (hunk 12 above)
Comment (alexyarats) [reply gen-py/tutorial/RacconChat-remote]: done

File: tutorial.thrift (hunk 13 above)
Comment (alexyarats) [reply tutorial.thrift]: done

File: tutorial.thrift (hunk 14 above)
Comment (alexyarats) [reply tutorial.thrift]: done

File: tutorial.thrift (hunk 15 above)
Comment (alexyarats) [reply tutorial.thrift]: done

File: tutorial.thrift (hunk 16 above)
Comment (alexyarats) [reply tutorial.thrift]: done

File: PythonClient.py (hunk 3 above)
Comment (alexyarats) [reply PythonClient.py]: don't understand how it works

Comment (denisyarats) [reply PythonClient.py]: Here is an example:

`parser = argparse.ArgumentParser(description='freq map')`
`parser.add_argument('--input', help='input file')`
`parser.add_argument('--output', help='output file')`
`args = parser.parse_args()`
`f = open(args.input, 'r')`

then you can call you binary like this
`
./binary --input name_of_input_file --output name_of_output_file
`

File: gen-cpp/CppRacconChat.h (hunk 5 above)
Comment (denisyarats) [reply gen-cpp/CppRacconChat.h]: what compiler do you use? llvv-g++?

File: gen-cpp/CppServer.cpp (hunk 9 above)
Comment (denisyarats) [reply gen-cpp/CppServer.cpp]: updated the comment

File: gen-cpp/CppRacconChat.h (hunk 5 above)
Comment (alexyarats) [reply gen-cpp/CppRacconChat.h]: yes

File: gen-cpp/CppServer.cpp (hunk 9 above)
Comment (alexyarats) [reply gen-cpp/CppServer.cpp]: done.

File: PythonClient.py (hunk 3 above)
Comment (alexyarats) [reply PythonClient.py]: There is a problem with thrift.

Comment (alexyarats) [reply PythonClient.py]: done

File: gen-cpp/CppRacconChat.h (hunk 7 above)
Comment (alexyarats) [reply gen-cpp/CppRacconChat.h]: done

**Comment by alexyarats** [comment]
RaccooChat Version 1.1.
Last changes:
  - test_gflag.cpp - c++ file for testing gflag and glog.

Client.py was divided on two separate files:
  - client.py - client;
  - RaccooChatClient.py - class for working with server;

**Review by denisyarats — APPROVED** [review]

**Pull request merged**